### ISSUE TITLE
Virtual clusters should have unique kafka cluster ids.

### ISSUE THREAD

**Issue by k-wall** [kroxylicious/kroxylicious#379]
**Is your feature request related to a problem? Please describe.**

Apache Kafka has the notion of `clusterId` which is a unique stable identifier that identifies the kafka cluster to the client.  This was introduced by [KIP-78](https://cwiki.apache.org/confluence/display/KAFKA/KIP-78%3A+Cluster+Id).

Currently kroxylicious virtual clusters relay the upstream's kafka cluster id back to the client unchanged.  If kroxylicious defines several virtual clusters targeting the same upstream cluster, all of those virtual clusters will present the same `clusterId`.
This behaviour is in conflict with KIP-78.  It could cause kafka clients or other tooling to malfunction. 

All virtual clusters must have a unique stable cluster id that is different to the target cluster's own clusterid.

**Describe the solution you'd like**

We could introduce a virtual cluster configuration item that lets the user specify the clusterid.  It would be a responsibility of the user to generate one. They could use `/bin/kafka-storage.sh random-uuid`.

Perhaps for convenience, if the virtual cluster's `clusterId` is not present, Kroxylicious could manufacture a stable id from other virtualcluster attributes (perhaps a value derived from the target cluster `clusterid` and the virtual cluster's bootstrap address).  If kroxylicious had such a feature, I think we'd label it `recommended not for production use`.

**Describe alternatives you've considered**

**Additional context**

**Comment by robobario** [comment]
The question I had around listeners was, if the proxied cluster offers PLAIN on 9092 and TLS on 9093, how would we model that in Kroxylicious so that Producers and Consumers can connect to the PLAIN listener or the TLS listener?

For a 3 node cluster would that be a VirtualCluster with 1 bootstrap port for PLAIN, 1 bootstrap port for TLS, 3 broker ports for PLAIN, 3 broker ports for TLS. Or could we model it as two VirtualClusters, one that connects to the PLAIN listener and one that connects to the TLS listener? If they were modelled as two VirtualClusters in kroxylicious configuration they would still have the same cluster id.

**Comment by k-wall** [comment]
Yes, that does make sense.  Sorry for my thickness during the call!

You are right, if a kroxylicious user wanted to do that, multiple virtual clusters is really the way. I don't think that's the right approach. I agree is multi listener inside the virtual cluster approach is correct, or even plain/TLS port unification (sniff the first bytes from the wire and make a determination either way).

**Comment by SamBarker** [comment]
Exposing two different virtual clusters for different auth mechanisms seems a reasonable thing to do in certain circumstance. Present plain as `Kafka-internal-x.corp.lan` and sasl as `broker-x.my-vanity-domain.com` multi tenant lite. In this scenario should both clusters share a cluster id?

One answer is no we are presenting distinct clusters. Alternatively as @tombentley outlined on the call for Kroxylicious to be transparently added and removed they have to share a cluster id (as that's what the physical cluster will supply).

**Comment by robobario** [comment]
Also, a case from the call was that while users are trialling Kroxylicious they might want to use the backing clusterId so that the clients see continuity as they switch to and from Kroxylicious. Maybe this could be an explicit choice, where a user must say:

```
virtualClusters:
  demo-plain:
    clusterId: 1234-123124-2131
```

or 

```
virtualClusters:
  demo-plain:
    useTargetClusterId: true
```

and we can warn them off `useTargetClusterId` in docs.

**Comment by k-wall** [comment]
I have a worry that users may forget to make the id unique when it ought to be.  For instance, it would be very easy to template kroxylicious config file and inadvertently reuse the same id over and over again.  It might not be discovered until clients or tooling start breaking, choking on different clusters with the same ids.

I think Kroxylicious default should be to insist on unique ids for the virtual clusters and insist they are different from the target clusters. There would an option to relax the restriction.

**Comment by SamBarker** [comment]
I was starting to think about a setting such as `useTargetClusterId` when I was writing my response. 

I'm getting a bit lost in the details. What are the actual requirements we are trying to satisfy/what problem are we trying to solve? 

As I see it there are two "types" of virtual cluster<sup>[1]</sup>. 
1. Transparent clusters - Kroxylicious is mapping the cluster one to one for network transit/monitoring reasons.
2. Pseudo clusters - Kroxylicious is pretending there is a different cluster there.

In 1. We want to forward the `clusterId` from from the underlying cluster. For 2. we want a unique ID across all the clusters involved. 

I'm not persuaded that in 2. the `clusterId` should be configuration, however it needs to be stable and the simplest path to a stable ID is allowing it as configuration.

[1] The naming sucks.

**Comment by robobario** [comment]
Oh and a note, I don't think we need to tackle the multiple listener thing yet. I imaging proxying a single listener is enough to get people trying Kroxylicious out and getting value.

**Comment by k-wall** [comment]
I'm going to try to summarise the requirements:

1. virtual clusters must have a `uid` - this is the uid that the virtual cluster will use in response to a [DescribeCluster](https://kafka.apache.org/protocol.html#The_Messages_DescribeCluster) request.
2. the uid for the virtual cluster can either be:
    * the uid of the target cluster.  A use-case: where a proxy layer is to be transparently interposed between client / cluster without disturbing the cluster's established identity.
    * another uid (different to the target cluster's).
3. there is no requirement for `uids` for different virtual clusters to be distinct from one and other.  This allows for the use-case where the user wishes to expose a single target cluster as if it were many.
